src/expint.jl

```diff
@@ -0,0 +1,331 @@
+    gammaabs, cfabs = abs(gammapart), abs(cfpart)
+    if gammaabs != Inf && gammaabs > 1.0 && gammaabs > cfabs
+        # significant gamma part, use this
+        return gammapart + cfpart, iters, true
+
+    else
+        return En_cf_nogamma(x, ν, niter)..., false
+    end
+end
+
+# Compute expint(ν, z₀+Δ) given start = expint(ν, z₀)
+function En_taylor(ν::Number, start::Number, z₀::Number, Δ::Number)
+    a = exp(z₀) * start
+    k, iters = 0, 0
+    asum = a
+    Δ_prod_fact = -Δ
+    
+    while iters < 100 # perhaps
+        a_pre = Δ_prod_fact + a*Δ*(ν - k - 1)/(k+1)
+        a = a_pre / z₀
+        asum_prev = asum
+        asum += a
+        
+        if abs(asum_prev - asum) < 1e-15
+            break
+        end
+        
+        Δ_prod_fact *= -Δ / (k+2)
+        
+        iters += 1
+        k += 1
+    end
+
+    res = exp(-z₀) * asum
+    return res
+end
+
+# series about origin
+# https://functions.wolfram.com/GammaBetaErf/ExpIntegralE/06/01/04/01/01/0003/
+function En_expand_origin(x::Number, ν::Number)
+    if isreal(ν) && floor(ν) == ν
+        # go to special case for integer ν
+        return En_expand_origin(x, Int(ν))
+    end
+    gammaterm = gamma(1 - ν)
+    if abs(gammaterm) != 0
+        # don't multiply by NaN
+        gammaterm *= x^(ν-1)
+    end
+    frac = 1
+    sumterm = frac / (1 - ν)
+    k, maxiter = 1, 100
+    while k < maxiter
+        frac *= -x / k
+        prev = sumterm
+        sumterm += frac / (k + 1 - ν)
+        if abs(sumterm - prev) < 1e-15 * abs(prev)
+            break
+        end
+        k += 1
+    end
+    
+    return gammaterm - sumterm
+end
+
+# series about the origin, special case for integer n
+function En_expand_origin(x::Number, n::Integer)
+    gammaterm = 1
+    for i = 1:n-1
+        gammaterm *= -x / i
+    end
+
+    gammaterm *= digamma(n) - log(x)
+    sumterm = n == 1 ? 0 : 1 / (1 - n)
+    frac = 1
+    k, maxiter = 1, 100
+    while k < maxiter
+        frac *= -x / k
+        # skip term with zero denominator
+        if k != n-1
+            prev = sumterm
+            sumterm += frac / (k + 1 - n)
+            if abs(sumterm - prev) < 1e-15 * abs(prev)
+                break
+            end
+        end
+        k += 1
+    end
+    return gammaterm - sumterm
+end
+
+# can find imaginary part of E_ν(x) for x on negative real axis analytically
+function imagbranchcut(x::Number, ν::Number)
+    a = real(x)
+    impart = π * im * exp(-π*im*ν) * a^(ν-1) / gamma(ν)
+    # exp(n*log(z) - loggamma(n))
+    return imag(impart) * im # get rid of any real error
+end
+
+const ORIGIN_EXPAND_THRESH = 3
+function En(ν::Number, x::Number, niter::Int=1000, debug=false)
+    if x == 0.0
+        if real(ν) > 0
+            return 1.0 / (ν - 1)
+        else
+            return Inf
+        end
+    end
+    if ν == 0
+        return exp(-x) / x
+    end
+    # asymptotic test
+    if exp(-x) / x == 0
+        return zero(x)
+    end
+
+    if abs(x) < ORIGIN_EXPAND_THRESH
```

Comment:
Is this only for double precision?

Reply:
Note that you can use `abs(real(z)) + abs(imag(z))` to avoid the sqrt

Reply:
you can do something like:
```jl
En_expand_origin_threshold(z) = 3 # default
En_expand_origin_threshold(z::Union{Float32,ComplexF32}) = 2 # different threshold in single precision
```
and then call `En_expand_origin_threshold(z)` in this check.